File: packages/m9m-cli/python/m9m/binary.py
"""Binary download and caching for the m9m Python SDK."""

import os
import platform
import shutil
import stat
import subprocess
import sys
import urllib.request
from pathlib import Path


_GITHUB_REPO = "neul-labs/m9m"
_CACHE_DIR = Path.home() / ".m9m" / "bin"


def _get_platform() -> tuple[str, str]:
    """Return (os, arch) identifiers matching GitHub release assets."""
    system = platform.system().lower()
    machine = platform.machine().lower()

    os_name = {"darwin": "darwin", "linux": "linux", "windows": "windows"}.get(
        system, system
    )

    arch_name = {
        "x86_64": "amd64",
        "amd64": "amd64",
        "aarch64": "arm64",
        "arm64": "arm64",
    }.get(machine, machine)

    return os_name, arch_name


def _latest_release_version() -> str:
    """Fetch the latest release tag from GitHub (e.g. 'v0.2.0')."""
    try:
        import urllib.request
        import json

        url = f"https://api.github.com/repos/{_GITHUB_REPO}/releases/latest"
        req = urllib.request.Request(url, headers={"Accept": "application/vnd.github.v3+json"})
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
            return data.get("tag_name", "v0.2.0")
    except Exception:
        return "v0.2.0"


def get_binary_path() -> str:
    """Return the path to the cached m9m binary, downloading if necessary."""
    os_name, arch_name = _get_platform()
    binary_name = "m9m.exe" if os_name == "windows" else "m9m"
    cached = _CACHE_DIR / binary_name

    if cached.exists():
        return str(cached)

    download_binary()
    return str(cached)


def download_binary(version: str | None = None) -> str:
    """Download the m9m binary for this platform and cache it."""
    os_name, arch_name = _get_platform()
    if version is None:
        version = _latest_release_version()
    if not version.startswith("v"):
        version = f"v{version}"

    binary_name = "m9m.exe" if os_name == "windows" else "m9m"
    artifact_name = f"m9m-{os_name}-{arch_name}"
    url = f"https://github.com/{_GITHUB_REPO}/releases/download/{version}/{artifact_name}"
    cached = _CACHE_DIR / binary_name

    _CACHE_DIR.mkdir(parents=True, exist_ok=True)

    # Download
    req = urllib.request.Request(url, headers={"User-Agent": "m9m-python-sdk/1.0"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        cached.write_bytes(resp.read())

    # Make executable (Unix)
    if os_name != "windows":
        st = os.stat(cached)
        os.chmod(cached, st.st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)

    return str(cached)

File: packages/m9m-cli/nodejs/src/binary.ts
/**
 * Binary download and caching for the m9m Node.js SDK.
 */

import * as fs from 'fs';
import * as https from 'https';
import * as os from 'os';
import * as path from 'path';

const GITHUB_REPO = 'neul-labs/m9m';

function getCacheDir(): string {
  const home = os.homedir();
  return path.join(home, '.m9m', 'bin');
}

function getPlatform(): { os: string; arch: string } {
  const platform = os.platform();
  const arch = os.arch();

  const osName: Record<string, string> = {
    darwin: 'darwin',
    linux: 'linux',
    win32: 'windows',
  };

  const archName: Record<string, string> = {
    x64: 'amd64',
    arm64: 'arm64',
  };

  return {
    os: osName[platform] || platform,
    arch: archName[arch] || arch,
  };
}

async function latestReleaseVersion(): Promise<string> {
  return new Promise((resolve) => {
    const req = https.get(
      `https://api.github.com/repos/${GITHUB_REPO}/releases/latest`,
      {
        headers: {
          'User-Agent': 'm9m-nodejs-sdk/1.0',
          Accept: 'application/vnd.github.v3+json',
        },
      },
      (res) => {
        let data = '';
        res.on('data', (chunk) => (data += chunk));
        res.on('end', () => {
          try {
            const json = JSON.parse(data);
            resolve(json.tag_name || 'v0.2.0');
          } catch {
            resolve('v0.2.0');
          }
        });
      }
    );
    req.on('error', () => resolve('v0.2.0'));
    req.setTimeout(10000, () => {
      req.destroy();
      resolve('v0.2.0');
    });
  });
}

function downloadFile(url: string, dest: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(dest);
    const req = https.get(url, { headers: { 'User-Agent': 'm9m-nodejs-sdk/1.0' } }, (res) => {
      if (res.statusCode === 302 || res.statusCode === 301) {
        if (res.headers.location) {
          downloadFile(res.headers.location, dest).then(resolve).catch(reject);
          return;
        }
      }
      if (res.statusCode !== 200) {
        reject(new Error(`Download failed: ${res.statusCode}`));
        return;
      }
      res.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve();
      });
    });
    req.on('error', reject);
    req.setTimeout(60000, () => {
      req.destroy();
      reject(new Error('Download timeout'));
    });
  });
}

export async function downloadBinary(version?: string): Promise<string> {
  const { os: osName, arch } = getPlatform();
  const ver = version || (await latestReleaseVersion());
  const artifactName = `m9m-${osName}-${arch}`;
  const url = `https://github.com/${GITHUB_REPO}/releases/download/${ver}/${artifactName}`;

  const cacheDir = getCacheDir();
  if (!fs.existsSync(cacheDir)) {
    fs.mkdirSync(cacheDir, { recursive: true });
  }

  const binaryName = osName === 'windows' ? 'm9m.exe' : 'm9m';
  const cached = path.join(cacheDir, binaryName);

  await downloadFile(url, cached);

  if (osName !== 'windows') {
    fs.chmodSync(cached, 0o755);
  }

  return cached;
}

export function getBinaryPath(): string {
  const cacheDir = getCacheDir();
  const { os: osName } = getPlatform();
  const binaryName = osName === 'windows' ? 'm9m.exe' : 'm9m';
  const cached = path.join(cacheDir, binaryName);

  if (fs.existsSync(cached)) {
    return cached;
  }

  throw new Error(
    'm9m binary not found. Run downloadBinary() first or ensure it is downloaded during install.'
  );
}
